### DISCUSSION TITLE
How can UDFs be so much faster?

### DISCUSSION THREAD

**Discussion by skinkie** [duckdb/duckdb#15184] [Q&A] [answered]
I am deserialising a large XML document into individual Python dataclasses. This happens in SAX-style. After deserialisation XML into a dataclass, this dataclass is serialised with pickle, for storage into the database for quick access. My source table contains pickled objects, which are written in a `BINARY` column.

I am maintaining a referencing table which contains pairs of related objects and their ids. Considering the cost of deserialisation I have two strategies of creating the referencing table:
1. After the XML deserialisation of a single object, but while deserialising the document, I have the XML object available as dataclass, I can do the reference-computation and write both the pickled dataclass as the referencing rows into the database, this causes 1 + N `INSERT`s into the database.
2. Alternatively, after the XML deserialisation of all objects and writing them individually into the database, I call a UDF function, which deserialises the pickle object (currently limited by the Python GIL) which then unfolds all the results into a temporary table. This is effectively a single query.

For "small" documents, the total time spend of method one is roughly twice of the second method, which is doing the serialisation, and extra deserialisation round. I understand that I am doing significantly less `INSERT INTO` queries (1 vs N)
For larger documents I am already seeing a 3500% speed increase of using the UDF. That means that the entire database must be deserialised (again) computed and the references are returned as via the UDF, with a bit of SQL magic split into the reference table.

@Tishj do you think I should report an issue that the individual writes have such enormous performance hit? Or should I just "be very happy" with this enormous performance boost of the user defined function?

```
real    143m50.750s
user    192m33.595s
sys     7m15.433s

-rw-r--r-- 1 skinkie skinkie 144M Dec  7 21:09 /tmp/keolis.duckdb

D select count(*) from referencing;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│    212123    │
└──────────────┘
D select count(*) from embedded;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│    62238     │
└──────────────┘


real    4m38.518s
user    4m52.549s
sys     0m6.704s



D select count(*) from referencing;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│    212123    │
└──────────────┘
D select count(*) from embedded;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│    62238     │
└──────────────┘
```

**EDIT:** There was a slight programming error.
```
real    16m18.497s
user    17m57.214s
sys     0m35.605s
```
So 3500% woulde be more like 400%.

**Comment by soerenwolfers** [comment, marked as answer]
I don't think this needs an issue: It is absolutely expected that iterated `INSERT`s are slow. That's why DuckDB has an Appender API, see https://duckdb.org/docs/data/appender.html This is not currently available for the Python client, but a UDF is a good substitute, as you've discovered.
